File: templates/scripts/epic-fold.js
#!/usr/bin/env bun
'use strict';
/**
 * EPIC-FOLD: Wave 8 — Consolidate and compact events when an epic completes.
 *
 * Exported API:
 *   detectCompletedEpics({ cwd? }) → string[]
 *   foldEpic({ epic, cwd? })       → boolean
 *
 * CLI:
 *   node epic-fold.js --detect
 *   node epic-fold.js --epic <name> [--cwd <path>]
 *
 * Rules:
 *   - Fail-open: any error → warning to stderr, returns false / empty array.
 *   - Idempotent: calling foldEpic twice is safe (skips if already CLOSE or
 *     epic.complete event already exists in log).
 *   - Node built-ins only. No npm deps.
 *
 * @version 1.0.0
 */

const fs = require('fs');
const path = require('path');
const { execFileSync } = require('child_process');

// ── Harness event emitter (optional — fail-open if missing) ──────────────────

let harnessEvent = null;
try {
  harnessEvent = require('./../hooks/_lib/harness-event.js');
} catch (_) {}

// ── Helpers ───────────────────────────────────────────────────────────────────

function readJsonFile(filePath) {
  try {
    if (!fs.existsSync(filePath)) return null;
    return JSON.parse(fs.readFileSync(filePath, 'utf8'));
  } catch (_) {
    return null;
  }
}

function writeJsonFile(filePath, obj) {
  try {
    const dir = path.dirname(filePath);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(filePath, JSON.stringify(obj, null, 2) + '\n', 'utf8');
    return true;
  } catch (err) {
    process.stderr.write(`[epic-fold] warn: could not write ${filePath}: ${err.message}\n`);
    return false;
  }
}

/**
 * Parse an NDJSON file synchronously. Invalid lines are skipped.
 */
function readEventsSync(filePath) {
  const out = [];
  try {
    if (!fs.existsSync(filePath)) return out;
    const raw = fs.readFileSync(filePath, 'utf8');
    for (const line of raw.split(/\r?\n/)) {
      const trimmed = line && line.trim();
      if (!trimmed) continue;
      try { out.push(JSON.parse(trimmed)); } catch (_) {}
    }
  } catch (_) {}
  return out;
}

function emitEvent(eventName, payload, opts) {
  if (!harnessEvent || typeof harnessEvent.emit !== 'function') return false;
  try {
    return harnessEvent.emit(eventName, payload, {
      cwd: opts.cwd,
      spec: opts.spec || undefined,
      actor: { kind: 'script', id: 'epic-fold' },
    });
  } catch (_) {
    return false;
  }
}

// ── detectCompletedEpics ──────────────────────────────────────────────────────

/**
 * Scan .pipeline-states/*.json and return epic names that are candidates for fold:
 *   - parent_spec === null (root spec)
 *   - children_specs.length > 0
 *   - ALL children have phase === "CLOSE"
 *   - Root itself NOT yet phase === "CLOSE"
 *
 * Specs without parent/children fields (pre-Wave 7) are ignored.
 *
 * @param {object} opts  { cwd? }
 * @returns {string[]}   Array of epic spec names ready to fold.
 */
function detectCompletedEpics(opts) {
  const options = opts || {};
  const cwd = typeof options.cwd === 'string' ? options.cwd : process.cwd();
  const statesDir = path.join(cwd, '.claude', '.pipeline-states');

  const candidates = [];
  try {
    if (!fs.existsSync(statesDir)) return candidates;

    const files = fs.readdirSync(statesDir).filter(f => f.endsWith('.json'));
    for (const file of files) {
      try {
        const filePath = path.join(statesDir, file);
        const state = readJsonFile(filePath);
        if (!state) continue;

        // Must be a root spec (Wave 7 field)
        if (state.parent_spec !== null && state.parent_spec !== undefined) continue;
        const children = Array.isArray(state.children_specs) ? state.children_specs : [];
        if (children.length === 0) continue;

        // Root must NOT already be CLOSE
        const rootPhase = String(state.phaseName || state.phase || '').toUpperCase();
        if (rootPhase === 'CLOSE') continue;

        // All children must be CLOSE
        let allClosed = true;
        for (const child of children) {
          const childFile = path.join(statesDir, child + '.json');
          const childState = readJsonFile(childFile);
          if (!childState) { allClosed = false; break; }
          const childPhase = String(childState.phaseName || childState.phase || '').toUpperCase();
          if (childPhase !== 'CLOSE') { allClosed = false; break; }
        }

        if (allClosed) {
          const specName = state.spec || file.replace(/\.json$/, '');
          candidates.push(specName);
        }
      } catch (_) {}
    }
  } catch (err) {
    process.stderr.write(`[epic-fold] warn: detectCompletedEpics error: ${err.message}\n`);
  }
  return candidates;
}

// ── foldEpic ──────────────────────────────────────────────────────────────────

/**
 * Consolidate an epic that has all children in CLOSE:
 *  1. Check idempotency (already folded? skip).
 *  2. Read all harness events for epic + children.
 *  3. Aggregate stats.
 *  4. Emit epic.complete event.
 *  5. Write epic-summary entry to knowledge.json.
 *  6. Transition root to phase: "CLOSE".
 *  7. Emit epic.fold tombstone (marks granular events as compactable).
 *
 * @param {object} opts  { epic, cwd? }
 * @returns {boolean}    true if fold succeeded (or was already done).
 */
function foldEpic(opts) {
  const options = opts || {};
  const epic = typeof options.epic === 'string' ? options.epic.trim() : '';
  const cwd = typeof options.cwd === 'string' ? options.cwd : process.cwd();

  if (!epic) {
    process.stderr.write('[epic-fold] warn: --epic is required\n');
    return false;
  }

  try {
    const statesDir = path.join(cwd, '.claude', '.pipeline-states');
    const epicFile = path.join(statesDir, epic + '.json');
    const epicState = readJsonFile(epicFile);
    const eventsFile = path.join(cwd, '.claude', '.harness', 'events.jsonl');

    // ── Guard: root must exist ───────────────────────────────────────────────
    if (!epicState) {
      process.stderr.write(`[epic-fold] warn: pipeline-state not found for epic "${epic}"\n`);
      return false;
    }

    const children = Array.isArray(epicState.children_specs) ? epicState.children_specs : [];

    // ── Idempotency check 1: root already CLOSE ──────────────────────────────
    const currentPhase = String(epicState.phaseName || epicState.phase || '').toUpperCase();
    if (currentPhase === 'CLOSE') {
      // Already folded — silently succeed
      return true;
    }

    // ── Idempotency check 2: epic.complete event already in log ─────────────
    const existingEvents = readEventsSync(eventsFile);
    const alreadyComplete = existingEvents.some(
      ev => ev.event === 'epic.complete' && ev.payload && ev.payload.epic === epic
    );
    if (alreadyComplete) {
      // Transition root state if it somehow wasn't updated
      const state = readJsonFile(epicFile) || epicState;
      state.phase = 'CLOSE';
      state.phaseName = 'CLOSE';
      writeJsonFile(epicFile, state);
      return true;
    }

    // ── Step 2: Aggregate events for epic + children ─────────────────────────
    const specSet = new Set([epic, ...children]);
    const epicEvents = existingEvents.filter(ev => ev.spec && specSet.has(ev.spec));

    let findingsCount = 0;
    let decisionsCount = 0;
    let lessonsCount = 0;
    let toolCallsTotal = 0;
    let agentsTotal = 0;
    let minTs = null;
    let maxTs = null;

    const findingEvents = [];

    for (const ev of epicEvents) {
      if (ev.ts) {
        if (!minTs || ev.ts < minTs) minTs = ev.ts;
        if (!maxTs || ev.ts > maxTs) maxTs = ev.ts;
      }
      switch (ev.event) {
        case 'finding':
          findingsCount++;
          findingEvents.push(ev);
          break;
        case 'decision':
          decisionsCount++;
          break;
        case 'lesson':
          lessonsCount++;
          break;
        case 'tool.use':
          toolCallsTotal++;
          break;
        case 'agent.start':
          agentsTotal++;
          break;
        default:
          break;
      }
    }

    // Duration in ms
    let durationMs = 0;
    if (minTs && maxTs) {
      try {
        durationMs = new Date(maxTs).getTime() - new Date(minTs).getTime();
        if (!Number.isFinite(durationMs) || durationMs < 0) durationMs = 0;
      } catch (_) { durationMs = 0; }
    }

    const startedAt = minTs || new Date().toISOString();
    const endedAt = maxTs || new Date().toISOString();

    // Top 3 findings by confidence
    findingEvents.sort((a, b) => {
      const ca = (a.payload && typeof a.payload.confidence === 'number') ? a.payload.confidence : 0;
      const cb = (b.payload && typeof b.payload.confidence === 'number') ? b.payload.confidence : 0;
      return cb - ca;
    });
    const top3Findings = findingEvents.slice(0, 3);

    // ── Step 3: Emit epic.complete ───────────────────────────────────────────
    const completePayload = {
      epic,
      children: [...children],
      findings_count: findingsCount,
      decisions_count: decisionsCount,
      lessons_count: lessonsCount,
      tool_calls_total: toolCallsTotal,
      agents_total: agentsTotal,
      duration_ms: durationMs,
      started_at: startedAt,
      ended_at: endedAt,
    };

    emitEvent('epic.complete', completePayload, { cwd, spec: epic });

    // ── Step 4: Write to knowledge.json ─────────────────────────────────────
    // Build content string from top findings + decision/lesson counts
    const findingLines = top3Findings.map((fev, i) => {
      const content = (fev.payload && fev.payload.content) || '';
      const conf = (fev.payload && typeof fev.payload.confidence === 'number')
        ? fev.payload.confidence.toFixed(2)
        : '?';
      return `${i + 1}. [conf=${conf}] ${content}`;
    });

    const contentParts = [];
    if (findingLines.length > 0) {
      contentParts.push('Top findings:\n' + findingLines.join('\n'));
    }
    contentParts.push(`Decisions: ${decisionsCount}`);
    contentParts.push(`Lessons: ${lessonsCount}`);

    const knowledgeEntry = {
      type: 'epic-summary',
      name: epic,
      description: `Epic concluded with ${children.length} child spec(s): ${children.join(', ')}`,
      source: 'epic-fold',
      tags: ['epic', 'summary'],
      confidence: 0.85,
      cwd,
      // Extra fields stored via content field as extended description
      content: contentParts.join('\n\n'),
      spec_children: [...children],
      concluded_at: endedAt,
    };

    writeKnowledgeEntry(knowledgeEntry, cwd);

    // ── Step 5: Transition root to CLOSE ─────────────────────────────────────
    const updatedState = Object.assign({}, epicState);
    updatedState.phase = 'CLOSE';
    updatedState.phaseName = 'CLOSE';
    updatedState.folded_at = new Date().toISOString();
    writeJsonFile(epicFile, updatedState);

    // ── Step 6: Emit epic.fold tombstone ─────────────────────────────────────
    emitEvent('epic.fold', {
      epic,
      compactable_specs: [epic, ...children],
      folded_at: new Date().toISOString(),
    }, { cwd, spec: epic });

    return true;
  } catch (err) {
    process.stderr.write(`[epic-fold] warn: foldEpic("${epic}") error: ${err.message}\n`);
    return false;
  }
}

// ── writeKnowledgeEntry: sync write using knowledge-update.js protocol ────────

/**
 * Write an epic-summary entry to knowledge.json.
 * Uses the same dedup logic as knowledge-update.js (same name+type = update).
 * Falls back to direct file write if knowledge-update.js is unavailable.
 */
function writeKnowledgeEntry(entry, cwd) {
  // Try via knowledge-update.js (spawning it synchronously via stdin pipe)
  // After delegating to knowledge-update.js, patch extra epic-specific fields
  // directly since knowledge-update.js only handles the standard schema.
  try {
    const scriptPath = path.join(__dirname, 'memory.js');
    if (fs.existsSync(scriptPath)) {
      const input = JSON.stringify(entry);
      execFileSync(process.execPath, [scriptPath, 'knowledge'], {
        input,
        stdio: ['pipe', 'pipe', 'pipe'],
        timeout: 5000,
      });
      // Patch epic-specific fields that knowledge-update.js doesn't store
      patchEpicFields(entry, cwd);
      return true;
    }
  } catch (_) {}

  // Fallback: write directly to knowledge.json with same dedup logic
  try {
    const kbPath = path.join(cwd, '.claude', 'knowledge.json');
    let kb = { version: 1, entries: [] };
    try {
      if (fs.existsSync(kbPath)) {
        kb = JSON.parse(fs.readFileSync(kbPath, 'utf8'));
        if (!kb.entries) kb.entries = [];
      }
    } catch (_) {
      kb = { version: 1, entries: [] };
    }

    const ts = new Date().toISOString();
    const existIdx = kb.entries.findIndex(e => e.name === entry.name && e.type === entry.type);
    if (existIdx >= 0) {
      const ex = kb.entries[existIdx];
      ex.description = entry.description;
      ex.source = entry.source;
      ex.tags = entry.tags || [];
      ex.updatedAt = ts;
      ex.lastSeen = ts;
      ex.occurrences = (ex.occurrences || 1) + 1;
      ex.confidence = Math.min(1.0, 0.3 + ex.occurrences * 0.1);
      // Store extra epic fields
      if (entry.content) ex.content = entry.content;
      if (entry.spec_children) ex.spec_children = entry.spec_children;
      if (entry.concluded_at) ex.concluded_at = entry.concluded_at;
    } else {
      kb.entries.push({
        id: `epic-summary-${Date.now()}`,
        type: entry.type,
        name: entry.name,
        description: entry.description,
        source: entry.source,
        tags: entry.tags || [],
        confidence: 0.85,
        occurrences: 1,
        createdAt: ts,
        updatedAt: ts,
        lastSeen: ts,
        content: entry.content || '',
        spec_children: entry.spec_children || [],
        concluded_at: entry.concluded_at || ts,
      });
    }

    const dir = path.dirname(kbPath);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(kbPath, JSON.stringify(kb, null, 2), 'utf8');
    return true;
  } catch (err) {
    process.stderr.write(`[epic-fold] warn: writeKnowledgeEntry error: ${err.message}\n`);
    return false;
  }
}

/**
 * Patch epic-specific fields (spec_children, concluded_at, content) onto the
 * knowledge.json entry after knowledge-update.js has written the base entry.
 * This is a no-op if the fields are already present.
 */
function patchEpicFields(entry, cwd) {
  try {
    const kbPath = path.join(cwd, '.claude', 'knowledge.json');
    if (!fs.existsSync(kbPath)) return;
    const kb = JSON.parse(fs.readFileSync(kbPath, 'utf8'));
    if (!Array.isArray(kb.entries)) return;
    const idx = kb.entries.findIndex(e => e.name === entry.name && e.type === entry.type);
    if (idx < 0) return;
    const ex = kb.entries[idx];
    if (entry.spec_children) ex.spec_children = entry.spec_children;
    if (entry.concluded_at) ex.concluded_at = entry.concluded_at;
    if (entry.content) ex.content = entry.content;
    if (typeof entry.confidence === 'number'
        && (typeof ex.confidence !== 'number' || ex.confidence < entry.confidence)) {
      ex.confidence = entry.confidence;
    }
    fs.writeFileSync(kbPath, JSON.stringify(kb, null, 2), 'utf8');
  } catch (_) {}
}

module.exports = { detectCompletedEpics, foldEpic };

// ── CLI ───────────────────────────────────────────────────────────────────────
if (require.main === module) {
  (function () {
    try {
      const args = process.argv.slice(2);

      function getArg(name) {
        const idx = args.indexOf('--' + name);
        return idx >= 0 ? args[idx + 1] : null;
      }
      function hasFlag(name) {
        return args.includes('--' + name);
      }

      const cwd = getArg('cwd') || process.cwd();

      if (hasFlag('detect')) {
        const epics = detectCompletedEpics({ cwd });
        process.stdout.write(JSON.stringify({ epics_ready: epics }, null, 2) + '\n');
        process.exit(0);
      }

      const epicArg = getArg('epic');
      if (epicArg) {
        const ok = foldEpic({ epic: epicArg, cwd });
        process.stdout.write(JSON.stringify({ ok, epic: epicArg }) + '\n');
        process.exit(0);
      }

      // No recognised flag
      process.stderr.write('Usage:\n');
      process.stderr.write('  node epic-fold.js --detect [--cwd <path>]\n');
      process.stderr.write('  node epic-fold.js --epic <name> [--cwd <path>]\n');
    } catch (err) {
      process.stderr.write(`[epic-fold] error: ${err.message}\n`);
    }
    process.exit(0);
  })();
}

File: apps/rt/src/run/epic_fold.rs
//! `mustard-rt run epic-fold` — consolidate and compact harness events when
//! an epic completes.
//!
//! - `--detect` scans `.pipeline-states/*.json` and lists root specs whose
//!   children are all in phase `CLOSE` (and the root itself is not).
//! - `--epic <name>` folds one such epic: aggregates events for the epic + its
//!   children, emits an `epic.complete` event, writes an `epic-summary`
//!   knowledge entry (markdown), transitions the root to `CLOSE`, and emits
//!   an `epic.fold` tombstone.
//!
//! W4C migration: event aggregation reads per-spec NDJSON via
//! [`mustard_core::EventReader::stream`]; the `epic-summary` knowledge entry
//! is written as `.claude/knowledge/epic-{epic}.md` via
//! [`mustard_core::atomic_md::MarkdownStore`].
//!
//! Fail-open and idempotent.

use crate::util::now_iso8601;
use mustard_core::atomic_md::frontmatter::Frontmatter;
use mustard_core::atomic_md::{MarkdownDoc, MarkdownStore};
use mustard_core::fs;
use mustard_core::model::event::HarnessEvent;
use mustard_core::ClaudePaths;
use serde_json::{json, Map, Value};
use std::path::Path;

/// Read a JSON file, returning `None` on any error.
fn read_json(path: &Path) -> Option<Value> {
    let text = fs::read_to_string(path).ok()?;
    serde_json::from_str(&text).ok()
}

/// Write a JSON value pretty-printed with a trailing newline. Fail-soft.
#[cfg(test)]
fn write_json(path: &Path, value: &Value) -> bool {
    if let Some(parent) = path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    match serde_json::to_string_pretty(value) {
        Ok(text) => std::fs::write(path, format!("{text}\n")).is_ok(),
        Err(_) => false,
    }
}

/// The uppercased phase of a pipeline-state object — derived from the
/// `pipeline.phase` event log keyed by the state's `spec` name. Falls back to
/// the legacy in-JSON `phase` field for backwards compatibility.
fn state_phase(state: &Value, cwd: &Path) -> String {
    if let Some(spec) = state.get("spec").and_then(Value::as_str) {
        if let Some(phase) = crate::run::emit_phase::last_phase_for_spec(cwd, spec) {
            return phase.to_uppercase();
        }
    }
    state
        .get("phase")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_uppercase()
}

/// Read every harness event for `spec` from its per-spec NDJSON sink.
fn read_events_for_spec(cwd: &Path, spec: &str) -> Vec<HarnessEvent> {
    let Ok(cp) = ClaudePaths::for_project(cwd) else {
        return Vec::new();
    };
    let Ok(sp) = cp.for_spec(spec) else {
        return Vec::new();
    };
    mustard_core::projection::read_harness_events_from_ndjson_dir(&sp.events_dir())
}

/// Append a harness event for the given epic via the NDJSON route. Best-effort.
fn emit_event(project_dir: &str, event: &str, payload: Value, spec: &str) {
    let ts = now_iso8601();
    let sid = crate::run::env::session_id();
    let kind = crate::run::event_route::classify_kind(event);
    let _ = crate::run::event_writer_ndjson::write_event_with_ts(
        Path::new(project_dir),
        Some(spec),
        None,
        &sid,
        event,
        kind,
        Some(0),
        Some(&sid),
        Some("epic-fold"),
        None,
        &payload,
        Some(&ts),
    );
}

/// Scan `.pipeline-states/*.json` for epics ready to fold.
fn detect_completed_epics(cwd: &Path) -> Vec<String> {
    let states_dir = ClaudePaths::for_project(cwd)
        .map(|p| p.pipeline_states_dir())
        .unwrap_or_else(|_| cwd.to_path_buf());
    let Ok(entries) = fs::read_dir(&states_dir) else {
        return Vec::new();
    };
    let mut candidates = Vec::new();
    for entry in entries {
        let name = entry.file_name.clone();
        if !name.ends_with(".json") {
            continue;
        }
        let Some(state) = read_json(&entry.path) else {
            continue;
        };
        match state.get("parent_spec") {
            None | Some(Value::Null) => {}
            Some(_) => continue,
        }
        let children: Vec<&str> = state
            .get("children_specs")
            .and_then(Value::as_array)
            .map(|a| a.iter().filter_map(Value::as_str).collect())
            .unwrap_or_default();
        if children.is_empty() {
            continue;
        }
        if state_phase(&state, cwd) == "CLOSE" {
            continue;
        }
        let all_closed = children.iter().all(|child| {
            let child_file = states_dir.join(format!("{child}.json"));
            read_json(&child_file)
                .is_some_and(|cs| state_phase(&cs, cwd) == "CLOSE")
        });
        if all_closed {
            let spec = state
                .get("spec")
                .and_then(Value::as_str)
                .map_or_else(|| name.trim_end_matches(".json").to_string(), str::to_string);
            candidates.push(spec);
        }
    }
    candidates
}

/// Write an `epic-summary` markdown file under `.claude/knowledge/`.
fn write_knowledge_entry(
    cwd: &Path,
    epic: &str,
    name: &str,
    description: &str,
    content: &str,
    children: &[String],
    concluded_at: &str,
) {
    let Ok(cp) = ClaudePaths::for_project(cwd) else {
        return;
    };
    let dir = cp.claude_dir().join("knowledge");
    if fs::create_dir_all(&dir).is_err() {
        return;
    }
    let dest = dir.join(format!("epic-{epic}.md"));
    let mut fm = Map::new();
    fm.insert("kind".into(), json!("epic-summary"));
    fm.insert("name".into(), json!(name));
    fm.insert("confidence".into(), json!(0.85));
    fm.insert("source".into(), json!("epic-fold"));
    fm.insert("concluded_at".into(), json!(concluded_at));
    fm.insert(
        "spec_children".into(),
        json!(children.iter().cloned().collect::<Vec<_>>()),
    );
    fm.insert("status".into(), json!("active"));
    let body = format!("{description}\n\n{content}\n");
    let doc = MarkdownDoc {
        path: dest.clone(),
        frontmatter: Some(Frontmatter(Value::Object(fm))),
        body,
    };
    let _ = MarkdownStore::write_atomic(&dest, &doc);
}

/// Fold an epic — returns `true` on success (or when already folded).
fn fold_epic(cwd: &Path, epic: &str) -> bool {
    if epic.is_empty() {
        eprintln!("[epic-fold] warn: --epic is required");
        return false;
    }
    let paths = ClaudePaths::for_project(cwd).ok();
    let states_dir = paths
        .as_ref()
        .map(ClaudePaths::pipeline_states_dir)
        .unwrap_or_else(|| cwd.to_path_buf());
    let epic_file = paths
        .as_ref()
        .map(|p| p.pipeline_state_file(epic))
        .unwrap_or_else(|| states_dir.join(format!("{epic}.json")));
    let Some(epic_state) = read_json(&epic_file) else {
        eprintln!("[epic-fold] warn: pipeline-state not found for epic \"{epic}\"");
        return false;
    };
    let children: Vec<String> = epic_state
        .get("children_specs")
        .and_then(Value::as_array)
        .map(|a| a.iter().filter_map(Value::as_str).map(str::to_string).collect())
        .unwrap_or_default();

    // Idempotency 1: root already CLOSE.
    if state_phase(&epic_state, cwd) == "CLOSE" {
        return true;
    }

    // Aggregate events for the epic + its children via per-spec NDJSON sinks.
    let mut all_events: Vec<HarnessEvent> = read_events_for_spec(cwd, epic);
    for child in &children {
        all_events.extend(read_events_for_spec(cwd, child));
    }

    // Idempotency 2: an `epic.complete` event already exists for this epic.
    let already_complete = all_events.iter().any(|e| {
        e.event == "epic.complete"
            && e.payload.get("epic").and_then(Value::as_str) == Some(epic)
    });
    if already_complete {
        emit_event(
            cwd.to_string_lossy().as_ref(),
            "pipeline.phase",
            json!({ "from": null, "to": "CLOSE" }),
            epic,
        );
        return true;
    }

    let spec_set: std::collections::BTreeSet<&str> = std::iter::once(epic)
        .chain(children.iter().map(String::as_str))
        .collect();
    let mut findings_count = 0usize;
    let mut decisions_count = 0usize;
    let mut lessons_count = 0usize;
    let mut tool_calls_total = 0usize;
    let mut agents_total = 0usize;
    let mut min_ts: Option<String> = None;
    let mut max_ts: Option<String> = None;
    let mut finding_events: Vec<&HarnessEvent> = Vec::new();

    for ev in &all_events {
        let Some(spec) = ev.spec.as_deref() else {
            continue;
        };
        if !spec_set.contains(spec) {
            continue;
        }
        if !ev.ts.is_empty() {
            if min_ts.as_deref().is_none_or(|m| ev.ts.as_str() < m) {
                min_ts = Some(ev.ts.clone());
            }
            if max_ts.as_deref().is_none_or(|m| ev.ts.as_str() > m) {
                max_ts = Some(ev.ts.clone());
            }
        }
        match ev.event.as_str() {
            "finding" => {
                findings_count += 1;
                finding_events.push(ev);
            }
            "decision" => decisions_count += 1,
            "lesson" => lessons_count += 1,
            "tool.use" => tool_calls_total += 1,
            "agent.start" => agents_total += 1,
            _ => {}
        }
    }

    let started_at = min_ts.clone().unwrap_or_else(now_iso8601);
    let ended_at = max_ts.clone().unwrap_or_else(now_iso8601);
    let duration_ms = match (
        min_ts.as_deref().and_then(crate::run::complete_spec::parse_iso_millis),
        max_ts.as_deref().and_then(crate::run::complete_spec::parse_iso_millis),
    ) {
        (Some(a), Some(b)) => (b - a).max(0),
        _ => 0,
    };

    finding_events.sort_by(|a, b| {
        let ca = a.payload.get("confidence").and_then(Value::as_f64).unwrap_or(0.0);
        let cb = b.payload.get("confidence").and_then(Value::as_f64).unwrap_or(0.0);
        cb.partial_cmp(&ca).unwrap_or(std::cmp::Ordering::Equal)
    });
    let top3: Vec<&&HarnessEvent> = finding_events.iter().take(3).collect();

    emit_event(
        cwd.to_string_lossy().as_ref(),
        "epic.complete",
        json!({
            "epic": epic,
            "children": children,
            "findings_count": findings_count,
            "decisions_count": decisions_count,
            "lessons_count": lessons_count,
            "tool_calls_total": tool_calls_total,
            "agents_total": agents_total,
            "duration_ms": duration_ms,
            "started_at": started_at,
            "ended_at": ended_at,
        }),
        epic,
    );

    let finding_lines: Vec<String> = top3
        .iter()
        .enumerate()
        .map(|(i, fev)| {
            let content = fev.payload.get("content").and_then(Value::as_str).unwrap_or("");
            let conf = fev
                .payload
                .get("confidence")
                .and_then(Value::as_f64)
                .map_or_else(|| "?".to_string(), |c| format!("{c:.2}"));
            format!("{}. [conf={conf}] {content}", i + 1)
        })
        .collect();
    let mut content_parts: Vec<String> = Vec::new();
    if !finding_lines.is_empty() {
        content_parts.push(format!("Top findings:\n{}", finding_lines.join("\n")));
    }
    content_parts.push(format!("Decisions: {decisions_count}"));
    content_parts.push(format!("Lessons: {lessons_count}"));

    write_knowledge_entry(
        cwd,
        epic,
        epic,
        &format!(
            "Epic concluded with {} child spec(s): {}",
            children.len(),
            children.join(", ")
        ),
        &content_parts.join("\n\n"),
        &children,
        &ended_at,
    );

    emit_event(
        cwd.to_string_lossy().as_ref(),
        "pipeline.phase",
        json!({ "from": null, "to": "CLOSE" }),
        epic,
    );

    let mut compactable = vec![epic.to_string()];
    compactable.extend(children.iter().cloned());
    emit_event(
        cwd.to_string_lossy().as_ref(),
        "epic.fold",
        json!({
            "epic": epic,
            "compactable_specs": compactable,
            "folded_at": now_iso8601(),
        }),
        epic,
    );
    true
}

/// Dispatch `mustard-rt run epic-fold`.
pub fn run(detect: bool, epic: Option<&str>) {
    let cwd = std::env::current_dir().unwrap_or_else(|_| Path::new(".").to_path_buf());
    if detect {
        let epics = detect_completed_epics(&cwd);
        let out = json!({ "epics_ready": epics });
        println!("{}", serde_json::to_string_pretty(&out).unwrap_or_else(|_| "{}".to_string()));
        return;
    }
    if let Some(epic) = epic {
        let ok = fold_epic(&cwd, epic);
        println!("{}", json!({ "ok": ok, "epic": epic }));
        return;
    }
    eprintln!("Usage:");
    eprintln!("  epic-fold --detect");
    eprintln!("  epic-fold --epic <name>");
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    fn write_state(states: &Path, name: &str, value: Value) {
        write_json(&states.join(format!("{name}.json")), &value);
    }

    #[test]
    fn detect_finds_epic_with_all_children_closed() {
        let dir = tempdir().unwrap();
        let states = dir.path().join(".claude").join(".pipeline-states");
        std::fs::create_dir_all(&states).unwrap();
        write_state(
            &states,
            "epic",
            json!({ "spec": "epic", "parent_spec": null, "children_specs": ["c1"], "phase": "EXECUTE" }),
        );
        write_state(&states, "c1", json!({ "spec": "c1", "phase": "CLOSE" }));
        assert_eq!(detect_completed_epics(dir.path()), vec!["epic".to_string()]);
    }

    #[test]
    fn detect_skips_when_a_child_is_not_closed() {
        let dir = tempdir().unwrap();
        let states = dir.path().join(".claude").join(".pipeline-states");
        std::fs::create_dir_all(&states).unwrap();
        write_state(
            &states,
            "epic",
            json!({ "spec": "epic", "parent_spec": null, "children_specs": ["c1"], "phase": "EXECUTE" }),
        );
        write_state(&states, "c1", json!({ "spec": "c1", "phase": "QA" }));
        assert!(detect_completed_epics(dir.path()).is_empty());
    }

    #[test]
    fn fold_missing_epic_returns_false() {
        let dir = tempdir().unwrap();
        assert!(!fold_epic(dir.path(), "ghost"));
    }

    #[test]
    fn state_phase_falls_back_to_json_field_when_no_events() {
        let dir = tempdir().unwrap();
        let state = serde_json::json!({ "spec": "epic-x", "phase": "EXECUTE" });
        assert_eq!(state_phase(&state, dir.path()), "EXECUTE");
    }
}
